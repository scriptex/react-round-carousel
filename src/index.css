.carousel {
	width: 13.125rem;
	height: 13.125rem;
	position: relative;
	margin: 0 auto;
	perspective: 62.5rem;
}

.carousel__container {
	width: 100%;
	height: 100%;
	position: absolute;
	transform-style: preserve-3d;
	transition: transform 1s;
}

.carousel__slide {
	position: absolute;
	width: 11.875rem;
	height: 11.875rem;
	left: 0.625rem;
	top: 0.625rem;
	border: 0.125rem solid;
	transition: transform 1s, opacity 1s;
}

.carousel__slide img {
	width: 100%;
	height: auto;
	display: block;
}

.carousel__slide-overlay {
	color: #fff;
	text-align: center;
	display: flex;
	flex-flow: row wrap;
	align-items: center;
	justify-content: center;
	align-content: center;
	position: absolute;
	top: 0;
	right: 0;
	bottom: 0;
	left: 0;
	opacity: 0;
	transition: opacity 0.35s ease-in-out;
	background-color: rgba(0, 0, 0, 0.7);
	user-select: none;
}

.carousel__slide-overlay span,
.carousel__slide-overlay strong {
	display: block;
	flex: 0 0 100%;
}

.carousel__slide:hover .carousel__slide-overlay {
	opacity: 1;
}

.carousel__controls {
	display: flex;
	flex-flow: row wrap;
	align-items: center;
	justify-content: center;
}

.carousel__control {
	font-size: 0;
	line-height: 0;
	text-indent: -100%;
	width: 2.5rem;
	height: 2.5rem;
	display: block;
	position: relative;
	border: 1px solid;
	margin: 0 2rem;
	background: rgba(0, 0, 0, 0.7);
	border-radius: 50%;
	box-shadow: none;
}

.carousel__control::before,
.carousel__control::after {
	content: '';
	display: block;
	position: absolute;
	top: 0;
	right: 0;
	bottom: 0;
	left: 0;
	margin: auto;
}

.carousel__control::before {
	width: 50%;
	height: 1px;
	background-color: #fff;
}

.carousel__control::after {
	content: '';
	width: 0.9375rem;
	height: 0.9375rem;
	display: inline-block;
	vertical-align: middle;
	border-color: #fff;
	border-style: solid;
	border-width: 0 0 1px 1px;
	margin: auto;
	transform: rotate(-135deg);
	transform-origin: 50% 50%;
	transition: all 0.35s ease-in-out;
}

.carousel__control--prev::after {
	transform: rotate(45deg);
}
