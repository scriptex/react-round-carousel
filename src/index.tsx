import TouchSweep from 'touchsweep';
import { v4 as uuid } from 'uuid';
import {
	FC,
	useRef,
	useMemo,
	useState,
	useEffect,
	forwardRef,
	useCallback,
	CSSProperties,
	useImperativeHandle
} from 'react';

export type CarouselItem = Readonly<{
	alt?: string;
	image: string;
	content: React.ReactNode;
	onClick?: () => void;
}>;

export type DecoratedCarouselItem = CarouselItem & Readonly<{ id: string }>;

export type CarouselProps = Readonly<{
	ref?: React.ForwardedRef<CarouselRef>;
	items: CarouselItem[];
	itemWidth?: number;
	showControls?: boolean;
	slideOnClick?: boolean;
	classNamePrefix?: string;
	nextButtonContent?: string | React.ReactNode;
	prevButtonContent?: string | React.ReactNode;
}>;

export type CarouselRef = Readonly<{
	next: () => void;
	prev: () => void;
	getItems: () => DecoratedCarouselItem[];
	getSelectedIndex: () => number;
	setSelectedIndex: (index: number) => void;
}>;

export const Carousel: FC<CarouselProps> = forwardRef(
	(
		{
			items,
			itemWidth = 210,
			showControls = true,
			slideOnClick = false,
			classNamePrefix = 'carousel',
			prevButtonContent = 'Previous',
			nextButtonContent = 'Next'
		}: CarouselProps,
		CarouselRef
	) => {
		const data: DecoratedCarouselItem[] = useMemo(
			() =>
				items.map(item => ({
					...item,
					...((item as unknown as DecoratedCarouselItem).id
						? ({} as unknown as DecoratedCarouselItem)
						: { id: uuid() })
				})),
			[items]
		);

		const len = useMemo(() => data.length, [data.length]);
		const theta = useMemo(() => 360 / len, [len]);

		const radius = useMemo(
			() => Math.round(itemWidth / 2 / Math.tan(Math.PI / len)),
			[itemWidth, len]
		);

		const ref = useRef<HTMLDivElement>(null);
		const [selectedIndex, setSelectedIndex] = useState(0);

		const getSlideStyle = useCallback(
			(index: number): CSSProperties => {
				const style: CSSProperties = {};

				if (index < len) {
					const cellAngle = theta * index;

					style.opacity = 1;
					style.transform = `rotateY(${cellAngle}deg) translateZ(${radius}px)`;
				} else {
					style.opacity = 0;
					style.transform = 'none';
				}

				return style;
			},
			[len, radius, theta]
		);

		const getItemStyle = useCallback((): CSSProperties => {
			const angle = theta * selectedIndex * -1;

			return {
				transform: `translateZ(${-1 * radius}px) rotateY(${angle}deg)`
			};
		}, [radius, selectedIndex, theta]);

		const getClassName = useCallback(
			(parts: string | string[]) =>
				Array.isArray(parts)
					? parts
							.map((part: string) => `${classNamePrefix}${part}`)
							.join(' ')
					: `${classNamePrefix}${parts}`,
			[classNamePrefix]
		);

		const prev = useCallback(
			() => setSelectedIndex(selectedIndex - 1),
			[selectedIndex]
		);

		const next = useCallback(
			() => setSelectedIndex(selectedIndex + 1),
			[selectedIndex]
		);

		useEffect(() => {
			const area = ref?.current;
			const touchsweep = new TouchSweep(area ?? undefined);

			area?.addEventListener('swipeleft', next);
			area?.addEventListener('swiperight', prev);

			return () => {
				touchsweep.unbind();

				area?.removeEventListener('swipeleft', next);
				area?.removeEventListener('swiperight', prev);
			};
		});

		useImperativeHandle(
			CarouselRef,
			(): CarouselRef => ({
				next,
				prev,
				getItems: () => data,
				getSelectedIndex: () => selectedIndex,
				setSelectedIndex: (index: number) => setSelectedIndex(index)
			})
		);

		return (
			<>
				<div className={getClassName('')} ref={ref}>
					<div
						className={getClassName('__container')}
						style={getItemStyle()}
					>
						{data.map(
							(item: DecoratedCarouselItem, index: number) => (
								<div
									key={item.id}
									style={getSlideStyle(index)}
									onClick={() => {
										if (item.onClick) item.onClick();

										if (slideOnClick)
											setSelectedIndex(index);
									}}
									className={getClassName('__slide')}
								>
									<img src={item.image} alt={item.alt} />

									<div
										className={getClassName(
											'__slide-overlay'
										)}
									>
										{item.content}
									</div>
								</div>
							)
						)}
					</div>
				</div>

				{showControls && (
					<div className={getClassName('__controls')}>
						<button
							className={getClassName([
								'__control',
								'__control--prev'
							])}
							onClick={prev}
						>
							{prevButtonContent}
						</button>

						<button
							className={getClassName([
								'__control',
								'__control--next'
							])}
							onClick={next}
						>
							{nextButtonContent}
						</button>
					</div>
				)}
			</>
		);
	}
);

export default Carousel;
