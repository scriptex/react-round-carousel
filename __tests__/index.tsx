import React, { createRef } from 'react';
import { render, waitFor, fireEvent } from '@testing-library/react';

import { items } from '../demo/mocks';
import { Carousel, CarouselRef } from '../src';

const carouselItems = items
	.map(item => ({
		...item,
		image: item.image.replace(/\?.*$/gm, '')
	}))
	.slice(0, 2);

describe('Accordion', () => {
	it('should render properly with ref prop', async () => {
		const ref = createRef<CarouselRef>();
		const { asFragment, getAllByRole } = await waitFor(() =>
			render(<Carousel ref={ref} items={carouselItems} />)
		);

		expect(asFragment()).toMatchSnapshot();

		const carousel = ref.current;

		expect(carousel).toBeDefined();
		expect(carousel?.next).toBeInstanceOf(Function);
		expect(carousel?.prev).toBeInstanceOf(Function);
		expect(carousel?.getItems).toBeInstanceOf(Function);
		expect(carousel?.getSelectedIndex).toBeInstanceOf(Function);
		expect(carousel?.setSelectedIndex).toBeInstanceOf(Function);

		expect(carousel?.getItems().length).toEqual(2);
		expect(carousel?.getSelectedIndex()).toEqual(0);

		const [prev, next] = getAllByRole('button');

		await waitFor(() => fireEvent(next, new Event('click')));

		expect(asFragment()).toMatchSnapshot();

		await waitFor(() => fireEvent(prev, new Event('click')));

		expect(asFragment()).toMatchSnapshot();
	});

	it('should render properly without ref props', () => {
		const { asFragment } = render(<Carousel items={carouselItems} />);

		expect(asFragment()).toMatchSnapshot();
	});
});
