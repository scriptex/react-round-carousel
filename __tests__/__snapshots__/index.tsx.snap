// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accordion should render properly with ref prop 1`] = `
<DocumentFragment>
  <div
    class="carousel"
  >
    <div
      class="carousel__container"
      style="transform: translateZ(0px) rotateY(0deg);"
    >
      <div
        class="carousel__slide"
        style="opacity: 1; transform: rotateY(0deg) translateZ(0px);"
      >
        <img
          alt="A random Unsplash photo"
          src="https://picsum.photos/210"
        />
        <div
          class="carousel__slide-overlay"
        >
          <div>
            <strong>
              Round Carousel
            </strong>
            <span>
              Slide number 1
            </span>
          </div>
        </div>
      </div>
      <div
        class="carousel__slide"
        style="opacity: 1; transform: rotateY(180deg) translateZ(0px);"
      >
        <img
          alt="A random Unsplash photo"
          src="https://picsum.photos/211"
        />
        <div
          class="carousel__slide-overlay"
        >
          <div>
            <strong>
              Round Carousel
            </strong>
            <span>
              Slide number 2
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="carousel__controls"
  >
    <button
      class="carousel__control carousel__control--prev"
    >
      Previous
    </button>
    <button
      class="carousel__control carousel__control--next"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;

exports[`Accordion should render properly with ref prop 2`] = `
<DocumentFragment>
  <div
    class="carousel"
  >
    <div
      class="carousel__container"
      style="transform: translateZ(0px) rotateY(0deg);"
    >
      <div
        class="carousel__slide"
        style="opacity: 1; transform: rotateY(0deg) translateZ(0px);"
      >
        <img
          alt="A random Unsplash photo"
          src="https://picsum.photos/210"
        />
        <div
          class="carousel__slide-overlay"
        >
          <div>
            <strong>
              Round Carousel
            </strong>
            <span>
              Slide number 1
            </span>
          </div>
        </div>
      </div>
      <div
        class="carousel__slide"
        style="opacity: 1; transform: rotateY(180deg) translateZ(0px);"
      >
        <img
          alt="A random Unsplash photo"
          src="https://picsum.photos/211"
        />
        <div
          class="carousel__slide-overlay"
        >
          <div>
            <strong>
              Round Carousel
            </strong>
            <span>
              Slide number 2
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="carousel__controls"
  >
    <button
      class="carousel__control carousel__control--prev"
    >
      Previous
    </button>
    <button
      class="carousel__control carousel__control--next"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;

exports[`Accordion should render properly with ref prop 3`] = `
<DocumentFragment>
  <div
    class="carousel"
  >
    <div
      class="carousel__container"
      style="transform: translateZ(0px) rotateY(0deg);"
    >
      <div
        class="carousel__slide"
        style="opacity: 1; transform: rotateY(0deg) translateZ(0px);"
      >
        <img
          alt="A random Unsplash photo"
          src="https://picsum.photos/210"
        />
        <div
          class="carousel__slide-overlay"
        >
          <div>
            <strong>
              Round Carousel
            </strong>
            <span>
              Slide number 1
            </span>
          </div>
        </div>
      </div>
      <div
        class="carousel__slide"
        style="opacity: 1; transform: rotateY(180deg) translateZ(0px);"
      >
        <img
          alt="A random Unsplash photo"
          src="https://picsum.photos/211"
        />
        <div
          class="carousel__slide-overlay"
        >
          <div>
            <strong>
              Round Carousel
            </strong>
            <span>
              Slide number 2
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="carousel__controls"
  >
    <button
      class="carousel__control carousel__control--prev"
    >
      Previous
    </button>
    <button
      class="carousel__control carousel__control--next"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;

exports[`Accordion should render properly without ref props 1`] = `
<DocumentFragment>
  <div
    class="carousel"
  >
    <div
      class="carousel__container"
      style="transform: translateZ(0px) rotateY(0deg);"
    >
      <div
        class="carousel__slide"
        style="opacity: 1; transform: rotateY(0deg) translateZ(0px);"
      >
        <img
          alt="A random Unsplash photo"
          src="https://picsum.photos/210"
        />
        <div
          class="carousel__slide-overlay"
        >
          <div>
            <strong>
              Round Carousel
            </strong>
            <span>
              Slide number 1
            </span>
          </div>
        </div>
      </div>
      <div
        class="carousel__slide"
        style="opacity: 1; transform: rotateY(180deg) translateZ(0px);"
      >
        <img
          alt="A random Unsplash photo"
          src="https://picsum.photos/211"
        />
        <div
          class="carousel__slide-overlay"
        >
          <div>
            <strong>
              Round Carousel
            </strong>
            <span>
              Slide number 2
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="carousel__controls"
  >
    <button
      class="carousel__control carousel__control--prev"
    >
      Previous
    </button>
    <button
      class="carousel__control carousel__control--next"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;
